stan/math/prim/meta/value_type.hpp

```diff
@@ -0,0 +1,63 @@
+#ifndef STAN_MATH_PRIM_META_VALUE_TYPE_HPP
+#define STAN_MATH_PRIM_META_VALUE_TYPE_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/meta/is_eigen.hpp>
+#include <stan/math/prim/meta/is_vector.hpp>
+#include <stan/math/prim/meta/value_type.hpp>
+#include <type_traits>
+#include <vector>
+
+namespace stan {
+
+/** \ingroup type_trait
+ * Primary template class for metaprogram to compute the type of
+ * values stored in a container.
+ *
+ * tparam T type of container.
+ */
+template <typename T, typename = void>
+struct value_type {
+  using type = typename std::decay_t<T>;
+};
+
+/** \ingroup type_trait
+ * Specialization for pointers returns the underlying value the pointer is
+ * pointing to.
+ */
+template <typename T>
+struct value_type<T, std::enable_if_t<std::is_pointer<T>::value>> {
+  using type = typename value_type<std::remove_pointer<T>>::type;
+};
+
+/** \ingroup type_trait
+ * Helper function for accessing underlying type.
+ */
+template <typename T>
+using value_type_t = typename value_type<T>::type;
+
+/** \ingroup type_trait
+ * Template metaprogram class to compute the type of values stored
+ * in a standard vector.
+ *
+ * @tparam T type of elements in standard vector.
+ */
+template <typename T>
+struct value_type<T, std::enable_if_t<is_std_vector<T>::value>> {
```

Comment:
The arr/meta/value_type previously used require generics here (require_std_vector_t). But value_type is used in require_generics.hpp and that made value_type a dependency of require_generics and vice_versa. I used the underneath type traits directly here to fix this. 

cc @SteveBronder

Reply:
@rok-cesnovar : same question here.

Reply:
I was just pointing this out as these 2 changes are the only 2 where I didnt simply move around code from other files. All the additions are just moves from their current respective /arr/ or /mat/. 

Here I had to actually make changes to make the flattening work.